dump.go

```diff
@@ -137,3 +110,59 @@ func display(o io.Writer, f io.Reader, fname string) error {
 	fmt.Fprintf(o, "data = %v\n", rv.Elem().Interface())
 	return nil
 }
+
+func sizeof(r io.ReaderAt) (int64, error) {
+	switch r := r.(type) {
+	case interface{ Stat() (os.FileInfo, error) }:
+		fi, err := r.Stat()
+		if err != nil {
+			return 0, err
+		}
+		return fi.Size(), nil
+	case io.Seeker:
+		pos, err := r.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, err
+		}
+		sz, err := r.Seek(0, io.SeekEnd)
+		if err != nil {
+			return 0, err
+		}
+		_, err = r.Seek(pos, io.SeekStart)
+		if err != nil {
+			return 0, err
+		}
+		return sz, nil
+	default:
+		return 0, fmt.Errorf("npyio: unsupported  reader: %T", r)
+	}
+}
+
+func UnzipNpz(r io.ReaderAt) ([]*zip.File, error) {
+	var (
+		err      error
+		zipMagic = [4]byte{'P', 'K', 3, 4}
+	)
+
+	// detect .npz files (check if we find a ZIP file magic header)
+	var hdr [6]byte
+	_, err = r.ReadAt(hdr[:], 0)
+	if err != nil {
+		return nil, fmt.Errorf("npyio: could not infer format: %w", err)
+	}
+
+	sz, err := sizeof(r)
+	if err != nil {
+		return nil, fmt.Errorf("npyio: could not infer file size: %w", err)
+	}
+
+	if bytes.Equal(zipMagic[:], hdr[:len(zipMagic)]) {
```

Comment:
I would revert the test and exit early (with the error message from line 167), to save on "horizontal real estate".